[30기 한영현] Fix: 팀원, 래영님 리뷰 반영

younghyunHan opened the pull request:
## :: 최근 작업 주제 (하나 이상의 주제를 선택해주세요.)
- [x] 기능 추가
- [ ] 리뷰 반영
- [ ] 리팩토링
- [ ] 버그 수정
- [ ] 컨벤션 수정

<br />

## :: 구현 목표 (해당 브랜치에서 구현하고자 하는 하나의 목표를 설정합니다.)
1. Login, Main 개인 페이지 추가
2. 공통 Nav 컴포넌트 추가
3. 공통 Common.css 내용 맞춤
4. reset.css 추가
5. variables.css 추가
6. Mission1 - 사용자 입력 데이터 저장 기능 구현 완료
7. Mission2 - 로그인 버튼 활성화 ( validation)
8. Mission3 - 댓글 기능
9. Mission4 - 댓글 컴포넌트화 + props로 전달
10. Mock data 여러 개 댓글 기능 구현
11. Mock data 여러 개 피드 구현
12. 백 엔드 와 서버 통신 기능 추가

<br />

## :: 구현 사항 설명 (작업한 내용을 상세하게 기록합니다.)
1. 컴포넌트 담은 js 파일 안 최상위 태그 className을 컴포넌트 명과 동일하게 변경함.
2. 맥북 해상도를 고려해서  width: 1000px -> 1200px로 수정함
3. css파일 nesting 함.
4. Mission 1, 2 - onchange 이벤트 발생 시 handleInput함수 실행. 입력한 Id, pw 값을 조건에 맞추어 validaiton 한 후, 조건에 부합할 경우 버튼 색 변화, 버튼 활성 ( javaScript addEventListener에 인자로 넣는 onclick 이벤트 등이 리액트 에서는 속성으로 들어감 )
5. Mission 3, 4 - map메서드를 활용해서 댓글 기능 구현, 단 리액트에서는 객체, 배열에서 setState를 할 시 이전 내역이 사라짐. 그래서 이전 내역을 먼저 spread문법으로 넣어준 후, 새로 입력 받은 댓글을 추가함. 그리고 comment컴포넌트를 분리 후, 구조분해 할당으로 state값을 넘겨줌.
6. 백엔드와 서버 통신을 위해 먼저 회원가입을 진행함. email, password, name, phone_number를 넘겨줘서 회원가입 진행.
그 후, 아이디, password를 state로 백 엔드에 넘겨준 후, 맞다면 알림창 뜬 후, main 페이지로 이동.
7. useNavigate를 통해 조건을 주면서 이동



<br />

## :: 성장 포인트 (해당 기능을 구현하며 고민했던 사항이나 새로 알게된 부분, 어려웠던 점 등을 작성합니다.)
1. reset.css, common.css 등 기본 세팅  맞추고 하는 것의 중요성을 깨달음.
2. 태그 셀렉터를 사용할 경우 reset.css 파일을 먼저 import하더라도, reset.css 파일이 적용되는 문제점이 있었음. id 셀렉터는 아주 고유한 요소에만 사용해야 하므로, className을 사용해야 함.
3. css는 import로 각 페이지별로 주는 것이 구분하기 편리함.
4. index.js에서 reset.css -> common.css순서로 import
5. css파일은 public폴더에 직접 접근할 수 없기에 assets폴더 생성
6. useEffect를 통해 SideEffect를 한번 실행, 여러번 실행, 조건 줄 때 실행 하는 법을 배움. 이를통해 사이드 이펙트가 렌더링 막는 문제, 항상 실행되는 문제를 해결함.

<br />

## :: 기타 질문 및 특이 사항

younghyunHan committed:
Add: Mission 1 - 사용자 입력 데이터 저장 기능 구현

younghyunHan committed:
Add : Mission 2 - 로그인 버튼 활성화 기능 구현

younghyunHan committed:
Add : Mission 3 - 댓글 기능 구현

younghyunHan committed:
Add : Mission 4 - 댓글 컴포넌트화, props로 데이터 전달 구현

younghyunHan committed:
Add: 댓글 Mock data 추가

younghyunHan committed:
피드 MockData 추가중

hyejin30 commented on src/pages/younghyunHan/Login/Login.js:
```diff
@@ -1,7 +1,85 @@
+          className="loginButton"
+          onClick={goToMain}
+          disabled={disabled}
+          style={style}
```
```suggestion
          className={`loginBtn ${
              idValue.indexOf('@') > -1 && pwValue.length >= 5
                ? 'loginBtnLive'
                : ''
            }`}
```
 className의 동적 사용을 적용하면 좋을 것 같아요
인라인 스타일의 우선 순위가 높아서, 팀원이 class를 수정했을 때
class가 적용이 안되는 문제가 있다고 해요. 

조건 연산자를 활용해서 className을 동적으로 쓸 수 있음
className={원래 클래스 +활성화 클래스 <-- 조건 연산자로 클래스 추가 여부 결정}

lang92 replied:
김혜진 멘토님 리뷰 반영하기

lang92 replied:
+삼항연산자의 조건은 위 리뷰처럼 변수로 만들어서 사용하기

hyejin30 commented on src/pages/younghyunHan/Login/Login.js:
```diff
@@ -1,7 +1,85 @@
-  return <></>;
+  const [state, setState] = useState('');
+  const [passwordValue, setPasswordValue] = useState('');
+  const [disabled, setDisabled] = useState(true);
+  const [style, setStyle] = useState({ opacity: 0.5, cursor: 'default' });
```
className 동적 사용을 하게 되면
style 스테이트가 필요없게 된다고 해요!
disabled도 state로 남겨둘지 고민해보라고 하십니다 (from 멘토님..)

lang92 replied:
disabled, style 등은 state로 만들 필요 없음.
[리액트로 사고하기](https://ko.reactjs.org/docs/thinking-in-react.html) 참고
1. 부모로부터 props를 통해 전달됩니까? 그러면 확실히 state가 아닙니다.
2. 시간이 지나도 변하지 않나요? 그러면 확실히 state가 아닙니다.
3. 컴포넌트 안의 다른 state나 props를 가지고 계산 가능한가요? 그렇다면 state가 아닙니다.

OwlCJ commented on src/pages/younghyunHan/Main/Article.js:
```diff
@@ -0,0 +1,91 @@
+import './Article.scss';
+import { useState, useEffect } from 'react';
+import Comment from './Comment';
```
import  react 를 최상단으로 올리면 좋지 않을까요??

hyejin30 commented on src/pages/younghyunHan/Login/Login.js:
```diff
@@ -1,7 +1,85 @@
+    handleButton();
+  }
+
+  const checkId = idValue => {
+    return idValue.includes('@') ? true : false;
+  };
+
+  const checkPw = pwValue => {
+    return pwValue.length >= 5 ? true : false;
+  };
+
+  const handleButton = () => {
+    const isValidId = checkId(state);
+    const isValidPw = checkPw(passwordValue);
+
+    if (isValidId && isValidPw) {
+      buttonColor(true);
+    } else {
+      buttonColor(false);
+    }
+  };
+
+  const buttonColor = btnValid => {
+    setDisabled(!btnValid);
+    setStyle({
+      opacity: btnValid ? 1 : 0.5,
+      cursor: btnValid ? 'pointer' : 'default',
+    });
+  };
```
className 동적 사용을 하게 되면
유효성 검사 관련 함수를 다 지워도 될 것 같습니다!

hyejin30 reviewed (commented):
영현님 분발하세요

OwlCJ commented on src/pages/younghyunHan/Login/Login.js:
```diff
@@ -1,7 +1,85 @@
 import './Login.scss';
+import { useNavigate } from 'react-router-dom';
+import { useState } from 'react';
```
```suggestion
import { useState } from 'react';
import { useNavigate } from 'react-router-dom';
import './Login.scss';
```
scss를 밑으로 내려주면 좋을거같네요

OwlCJ commented on src/pages/younghyunHan/Login/Login.js:
```diff
@@ -1,7 +1,85 @@
+    });
+  };
+
+  const navigate = useNavigate();
```
훅 밑으로 붙여주는게 좋지 않을까요?

hyejin30 commented on src/pages/younghyunHan/Main/Main.js:
```diff
@@ -1,8 +1,158 @@
+          {articles.map((content, index) => {
+            return (
+              <div key={index}>
+                <Article
```
컴포넌트 이름을 Feed 로 바꿔보면 좋을 것 같아요!!
이름에 의미를 더 담아주면 좋을 것 같습니다

younghyunHan replied:
feedcontainer 안 feed

younghyunHan replied:
main-right position pixed

hyejin30 commented on src/pages/younghyunHan/Main/Main.js:
```diff
@@ -1,8 +1,158 @@
+            );
+          })}
+        </div>
+        <div className="mainRight">
```
mainRight도 컴포넌트로 만들어주면 좋을 것 같아요
이렇게 해주면 main.js 내용, scss 내용도 더 간단해질 것 같음!
```suggestion
        <MainRight />
```

hyejin30 commented on src/pages/younghyunHan/Main/Main.js:
```diff
@@ -1,8 +1,158 @@
+              </div>
+            </div>
+          </div>
+          <div className="companyInformation">
```
```suggestion
          <Footer />
```
컴포넌트로 분리해주고, footer 시멘틱 태그를 사용해보면 좋을 듯!

hyejin30 commented on src/pages/younghyunHan/Main/Main.js:
```diff
@@ -1,8 +1,158 @@
+            </div>
+          </div>
+          <div className="companyInformation">
+            <div className="companyInformationOne">
+              Instagram 정보 · 지원 · 홍보 센터 · API ·
+            </div>
+            <div className="companyInformationTwo">
+              채용 정보 · 개인정보처리방침 · 약관 ·
+            </div>
+            <div className="companyInformationThree">
+              디렉토리 · 프로필 · 해시태그 · 언어
+            </div>
+            <div className="companyInformationFour">2022 INSTAGRAM</div>
+          </div>
```
각 메뉴를 JS 파일에 저장하고,
map 메소드로 출력해주는 방식으로 수정해보면 좋을 듯해용

OwlCJ commented on src/pages/younghyunHan/Main/Main.scss:
```diff
@@ -0,0 +1,357 @@
+        }
+
+        .friendStoryNumber::-webkit-scrollbar {
+          display: none;
```
이런 선택자 활용으로 스크롤 감추신 부분 좋네요..

OwlCJ reviewed (commented):
아주 좋네요..

younghyunHan committed:
Add: mock data를 활용하여 여러 개의 피드 구현

younghyunHan committed:
fix:댓글 게시 기능 수정
- 댓글 입력하고 enter시 입력하는 사람 이름이 나오지 않던 부분 상수 데이터로 해결함.
- 상수 데이터 ( userName="test01" )

younghyunHan committed:
Add : 백엔드와 서버 통신 기능 추가

younghyunHan committed:
Add: main.js 댓글 컴포넌트 index 삭제

younghyunHan committed:
Add: Footer태그 사용해서 컴포넌트 분리

younghyunHan committed:
MainRight 컴포넌트 생성, Article컴포넌트 -> Feed컴포넌트로 이름 변경

lang92 commented on public/index.html:
```diff
@@ -8,6 +8,18 @@
+      href="https://use.fontawesome.com/releases/v5.14.0/css/all.css"
+      integrity="sha384-HzLeBuhoNPvSl5KYnjx0BT+WB0QEEqLprO+NBkkk5gbc67FTaL7XIGa2w1L0Xbgc"
+      crossorigin="anonymous"
+    />
```
공통 파일

lang92 commented on src/components/Nav/Nav.js:
```diff
@@ -1,7 +1,49 @@
 
 function Nav() {
-  return <></>;
+  return (
```
공통 파일

lang92 commented on src/pages/younghyunHan/Login/Login.js:
```diff
@@ -1,7 +1,104 @@
 
 function Login() {
-  return <></>;
+  const [state, setState] = useState('');
```
state라는 명칭이 무슨 데이터를 관리하는지 알 수 없음. 명확한 명칭으로 수정

lang92 commented on src/pages/younghyunHan/Login/Login.js:
```diff
@@ -1,7 +1,104 @@
+
+  const checkPw = pwValue => {
+    return pwValue.length >= 5 ? true : false;
+  };
```
- 기능을 함수로 분리하는 부분은 좋음
- 삼항연산자로 사용할 필요 없이, 조건 자체로 boolean이 될 수 있음
```suggestion
  const checkId = idValue => {
    return idValue.includes('@');
  };

  const checkPw = pwValue => {
    return pwValue.length >= 5;
  };
```
- 추가적으로,
```js
const isInputValid = idValue.includes('@') && pwValue.length >= 5
```
이런 식으로 한꺼번에 변수로 만들어 처리할 수도 있음

lang92 commented on src/pages/younghyunHan/Login/Login.js:
```diff
@@ -1,7 +1,104 @@
+
+  const navigate = useNavigate();
+
+  function handleIdInput(event) {
```
명칭만 봤을 때는 id 부분만 관리하는 함수인 것 같은데, 동작을 보면 아님. 수정

lang92 commented on src/pages/younghyunHan/Login/Login.js:
```diff
@@ -1,7 +1,104 @@
+      setState(event.target.value);
+    } else if (event.target.type === 'password') {
+      setPasswordValue(event.target.value);
+    }
```
구조분해할당!
```suggestion
    const { type, value } = event.target;
    
    if (type === 'text') {
      setState(value);
    } else if (type === 'password') {
      setPasswordValue(value);
    }
```

lang92 commented on src/pages/younghyunHan/Login/Login.js:
```diff
@@ -1,7 +1,104 @@
+      buttonColor(true);
+    } else {
+      buttonColor(false);
+    }
```
JS로 구현하던 방식 그대로 하셨는데, 기존 state로 계산이 가능하기 때문에 불필요한 state

lang92 commented on src/pages/younghyunHan/Login/Login.js:
```diff
@@ -1,7 +1,104 @@
+      body: JSON.stringify({
+        email: state,
+        password: passwordValue,
+        // name: '래영',
+        // phone_number: '010-1111-111',
```
ㅋㅋㅋㅋㅋㅋㅋ

lang92 commented on src/pages/younghyunHan/Main/Comment.js:
```diff
@@ -0,0 +1,14 @@
+import './Comment.scss';
+
+function Comment({ userName, content }) {
```
매개변수 자리에서 구조분해할당 굳

lang92 commented on src/pages/younghyunHan/Main/Feed.js:
```diff
@@ -0,0 +1,93 @@
+  const [cmtContLists, setCmtContLists] = useState([]);
+
+  const onKeyPress = event => {
+    let nextContents = event.target.value;
```
- 재할당되지 않는 변수는 const로 선언하기
- 이 부분은 변수 선언보단 구조분해할당 적용하기

lang92 commented on src/pages/younghyunHan/Main/Feed.js:
```diff
@@ -0,0 +1,93 @@
+import Comment from './Comment';
+
+function Feed(props) {
+  const [cmtContents, setCmtContents] = useState([]);
```
- input의 value가 들어오는 부분인데, 초기값이 string이 아닌 배열이 들어있음
- cmtContents라는 변수명과 배열인 초기값의 조합이 배열 데이터를 연상시킴
- 명확하지만 간결한 state명으로 수정. ex) `comment` / `commentList`

lang92 commented on src/pages/younghyunHan/Main/Feed.js:
```diff
@@ -0,0 +1,93 @@
+  const [cmtContents, setCmtContents] = useState([]);
+  const [cmtContLists, setCmtContLists] = useState([]);
+
+  const onKeyPress = event => {
```
함수명 수정!
- comment를 추가하는 역할을 하는 함수
- 엔터키를 쳤을 때 위 함수를 실행시키는 함수

로 분할하면 좋을 듯

lang92 commented on src/pages/younghyunHan/Main/Feed.js:
```diff
@@ -0,0 +1,93 @@
+    <div className="feed">
+      <article>
+        <div className="feedTop">
+          <div className="feedTopOne">
```
동일한 스타일을 가지는 부분인데, one two three 이런 식으로 나타낼 필요 없이 내용을 포괄하는 명칭의 동일한 className으로 사용

lang92 commented on src/pages/younghyunHan/Main/Feed.scss:
```diff
@@ -0,0 +1,132 @@
+article {
```
- article 태그가 아니라 feed라는 className으로 nesting하기
- 최상위 div를 article로 바꾸고 1 depth 줄이면 될 듯

lang92 commented on src/pages/younghyunHan/Main/Footer.scss:
```diff
@@ -0,0 +1,23 @@
+
+  .companyInformationFour {
+    color: #a3a49a;
+  }
```
위 리뷰 참고

lang92 commented on src/pages/younghyunHan/Main/Main.js:
```diff
@@ -1,8 +1,43 @@
 import '../../../styles/variables.scss';
+import Feed from './Feed';
+import MainRight from './MainRight';
+import { useState, useEffect } from 'react';
```
import 순서 컨벤션에 맞게 수정

younghyunHan committed:
Add: footer 상수 데이터 컴포넌트화
- 배열, 객체 활용해서 상수데이터 js파일 생성
- footer.js에서 Map메서드 활용해 상수데이터 출력
- 최상위 태그로 묶어줘야 함을 배움

lang92 reviewed (commented):
영현님 리뷰 확인해주세요!

younghyunHan committed:
Fix: 래영님, 팀원 리뷰 반영
